qiskit/opflow/gradients/natural_gradient.py

```diff
@@ -120,33 +120,65 @@ def convert(
-            c = np.real(x[0])
-            a = np.real(x[1])
-            if self.regularization:
-                # If a regularization method is chosen then use a regularized solver to
-                # construct the natural gradient.
-                nat_grad = NaturalGradient._regularized_sle_solver(
-                    a, c, regularization=self.regularization
-                )
-            else:
-                try:
-                    # Try to solve the system of linear equations Ax = C.
-                    nat_grad = np.linalg.solve(a, c)
-                except np.linalg.LinAlgError:  # singular matrix
-                    nat_grad = np.linalg.lstsq(a, c)[0]
-            return np.real(nat_grad)
+            return self.nat_grad_combo_fn(x, self.regularization)
 
         # Define the ListOp which combines the gradient and the QFI according to the combination
         # function defined above.
         return ListOp([grad, metric], combo_fn=combo_fn)
 
+    @staticmethod
+    def nat_grad_combo_fn(x: tuple, regularization: Optional[str] = None) -> np.ndarray:
+        r"""
+        Natural Gradient Function Implementation.
+
+        Args:
+            x: Iterable consisting of Gradient, Quantum Geometric Tensor.
+            regularization: Regularization method.
+
+        Returns:
+            Natural Gradient.
+
+        Raises:
+            ValueError: If the gradient has imaginary components that are non-negligible.
+
+        """
+        gradient = x[0]
+        metric = x[1]
+        if any(np.abs(np.imag(c_item)) > 1e-8 for c_item in gradient):
+            raise ValueError("The imaginary part of the gradient are non-negligible.")
+        gradient = np.real(gradient)
+
+        if regularization is not None:
+            # If a regularization method is chosen then use a regularized solver to
+            # construct the natural gradient.
+            nat_grad = NaturalGradient._regularized_sle_solver(
+                metric, gradient, regularization=regularization
+            )
+        else:
+            # Check if numerical instabilities lead to a metric which is not positive semidefinite
+            w, v = np.linalg.eigh(metric)
+
+            if not all(ew >= -1e-8 for ew in w):
+                raise ValueError(
+                    "The underlying metric has ein Eigenvalue < -1e-8. "
+                    "Please use a regularized least-square solver for this problem or"
+                    "increase the number of backend shots.",
+                )
+            if not all(ew >= 0 for ew in w):
+                # If not all eigenvalues are non-negative, set them to a small positive
+                # value
+                w = [max(1e-10, ew) for ew in w]
+                # Recompose the adapted eigenvalues with the eigenvectors to get a new metric
+                metric = np.real(v @ np.diag(w) @ np.linalg.inv(v))
+            nat_grad = np.linalg.lstsq(metric, gradient, rcond=1e-2)[0]
```

Comment:
Can we avoid hardcoding `1e-2`?

Reply:
We can. But this shouldn't be part of this PR. But a later PR where a custom solver can be handed to a NaturalGradient object.